tlsn/tlsn-core/tests/api.rs

```diff
@@ -0,0 +1,232 @@
+            *decommit2.nonce(),
+        ),
+    ];
+
+    // At the end of the session the User holds these artifacts:
+
+    // time when the TLS handshake began
+    let time = testdata.time;
+
+    // merkle tree of all User's commitments (the root of the tree was sent to the Notary earlier)
+    let merkle_tree = MerkleTree::from_leaves(&[commit1, commit2]).unwrap();
+
+    // encoder seed revealed by the Notary at the end of the label commitment protocol
+    let encoder_seed: [u8; 32] = notary_encoder_seed;
+
+    // server ephemeral key (known both to the User and the Notary)
+    let ephem_key = testdata.pubkey.clone();
+
+    // handshake data (to which the User sent a commitment earlier)
+    let handshake_data = HandshakeData::new(
+        ServerCertDetails::new(
+            vec![
+                testdata.ee.clone(),
+                testdata.inter.clone(),
+                testdata.ca.clone(),
+            ],
+            vec![],
+            None,
+        ),
+        ServerKxDetails::new(
+            testdata.kx_params(),
+            DigitallySignedStruct::new(SignatureScheme::RSA_PKCS1_SHA256, testdata.sig.clone()),
+        ),
+        testdata.cr,
+        testdata.sr,
+    );
+
+    // Commitment to the handshake which the User sent at the start of the TLS handshake
+    let (hs_decommitment, hs_commitment) = handshake_data.hash_commit();
+
+    let artifacts = SessionArtifacts::new(
+        time,
+        merkle_tree.clone(),
+        encoder_seed,
+        ephem_key.clone(),
+        hs_decommitment,
+    );
+
+    // Some outer context generates an (ephemeral) signing key for the Notary, e.g.
+    let rng = ChaCha20Rng::from_seed([6u8; 32]);
+    let signing_key = SigningKey::random(rng);
+    let raw_key = signing_key.to_bytes();
+
+    // Notary receives the raw signing key from some outer context
+    let mut signer = SigningKey::from_bytes(&raw_key).unwrap();
+    let notary_pubkey = signer.verifying_key();
+
+    // Notary creates the session header
+    assert!(data_sent.len() <= (u32::MAX as usize) && data_recv.len() <= (u32::MAX as usize));
+
+    let header = SessionHeader::new(
+        notary_encoder_seed,
+        merkle_tree.root(),
+        data_sent.len() as u32,
+        data_recv.len() as u32,
+        // the session's end time and TLS handshake start time may be a few mins apart
+        HandshakeSummary::new(time + 60, ephem_key, hs_commitment),
+    );
+
+    let signature = signer.sign(&header.to_bytes());
+    // Notary creates a msg and sends it to User
+    let msg = SignedSessionHeader {
+        header,
+        signature: signature.into(),
+    };
+
+    //---------------------------------------
+    let msg_bytes = bincode::serialize(&msg).unwrap();
+    let SignedSessionHeader { header, signature } = bincode::deserialize(&msg_bytes).unwrap();
+    //---------------------------------------
+
+    // User verifies the header and stores it with the signature in NotarizedSession
```

Comment:
I cant see the step where the User verifies the Notary's signature. I had it in my original test.

Reply:
One reason is consistency of the approach: we chose the approach of the User validating the data at the time when it is stored. 
Also if a malicious Notary provides the wrong sig and the User doesn't detect the wrong sig immediately, it may disrupt the User who counts on being able to create a proof later.